<launch>

  <arg name="uav_name" default="$(optenv UAV_NAME uav3)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="calib" default="false"/>

  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="true"/>

  <arg name="frame_uvdar" default="uvcam_$(arg uav_name)"/>
  <arg name="frame_camera" default="color_camera"/>


  <arg name="offline" default="false"/>


  <arg name="threshold" default="180"/>

  <arg name="publish_visualization"         default="false"/>
  <arg name="visualization_rate"         default="1"/>

  
 

  <arg name="device" default="$(optenv BLUEFOX_UV)"/>
  <arg name="rate" default="100"/>
  <arg name="serial" default="$(arg device)"/>
  <arg name="camera_name" default="mv_$(arg device)"/>
  <arg name="camera" default="$(arg camera_name)"/>
  <arg name="frame_id" default="$(arg camera)"/>
  <arg name="calib_url" default="file://$(find mrs_general)/camera_info/uvdar_bluefox.yaml"/>
  <arg name="fps" default="$(arg rate)"/>
  <arg name="idpf" default="0"/>
  <arg name="aec" default="false"/>
  <arg name="expose_us" default="1500"/>
  <arg name="agc" default="false"/>
  <arg name="gain_db" default="0.0"/>
  <arg name="cbm" default="0"/>
  <arg name="ctm" default="1"/>
  <arg name="dcfm" default="0"/>
  <arg name="hdr" default="false"/>
  <arg name="wbp" default="-1"/>
  <arg name="request" default="0"/>
  <arg name="mm" default="0"/>
  <arg name="jpeg_quality" default="90"/>

  <arg name="blink_process_rate" default="20"/>

  <arg name="accumulator_length"         default="23"/>
  <arg name="pitch_steps"         default="16"/>
  <arg name="yaw_steps"         default="16"/>
  <arg name="max_pixel_shift"         default="2"/>
  <arg name="reasonable_radius"         default="6"/>
  <arg name="nullify_radius"         default="8"/>

  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="view" default="false"/>

  <!-- <arg name="gui" default="false"/> -->

  <arg name="trajectoryControl" default="true"/>

  <arg name="followDistance" default="5"/>
  <arg name="trajCoeff" default="0.8"/>
  <arg name="yawCoeff" default="0.1"/>
  <arg name="tailingCoeff" default="1.0"/>

  <arg name="SamplePointSize" default="8"/>
  <arg name="StepSize" default="10"/>
  <arg name="publish" default="true"/>
  <arg name="cemeraRotated" default="false"/>
  <arg name="FromVideo" default="false"/>
  <arg name="VideoNumber" default="9"/>
  <arg name="surroundRadius" default="3"/>
  <arg name="camNum" default="0"/>
  <arg name="Delay" default="100"/>
  <arg name="TimeScale" default="1.0"/>

  <arg name="calib_file_uv" default="calib_results_bf_uv_fe.txt"/>

  <arg name="mask_file" default="mask_1.bmp"/>

  <arg name="frequency1"  default="6"/>
  <arg name="frequency2"  default="10"/>
  <arg name="frequency3"  default="15"/>
  <arg name="frequency4"  default="30"/>
      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <node
    name="body_uvcam_tf"
    pkg="tf"
    type="static_transform_publisher"
    args="0.09 0.04 0.2 -1.57079632679 0.00000000000 -1.57079632679 $(arg uav_name)/fcu uvcam_$(arg uav_name) 100"/>

  <group ns="$(arg uav_name)">
    <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen">
    <!-- <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen" launch-prefix="debug_roslaunch"> -->
      <param name="num_worker_threads" value="8" />
    </node>
  
    <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="bash -c 'sleep 5; $0 $@'"> 
        <!-- <param name="debug"		type="bool" value="true"/> --> 
        <param name="debug"		type="bool" value="$(arg debug)"/> 
        <param name="visual_debug"		type="bool" value="$(arg visual_debug)"/> 
        <param name="gui"             type="bool" value="$(arg gui)"/>
        <param name="InvertedPoints"    type="bool" value="false"/>

      <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3), $(arg frequency4)] </rosparam>
  

        <param name="publish_visualization"         type="bool" value="$(arg publish_visualization)"/>
        <param name="visualization_rate"         type="int" value="$(arg visualization_rate)"/>
 

      <param name="blink_process_rate"         type="int" value="$(arg blink_process_rate)"/>

      <param name="accumulator_length"         type="int" value="$(arg accumulator_length)"/>
      <param name="pitch_steps"         type="int" value="$(arg pitch_steps)"/>
      <param name="yaw_steps"         type="int" value="$(arg yaw_steps)"/>
      <param name="max_pixel_shift"         type="int" value="$(arg max_pixel_shift)"/>
      <param name="reasonable_radius"         type="int" value="$(arg reasonable_radius)"/>
      <param name="nullify_radius"         type="int" value="$(arg nullify_radius)"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>


      <rosparam param="camera_topics"> ["camera"] </rosparam>
      <rosparam param="points_seen_topics"> ["pointsSeen"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkersSeen"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimatedFramerate"] </rosparam>

      <remap from="~camera" to="/$(arg uav_name)/uvdar_bluefox/image_raw"/>
      <remap from="~pointsSeen" to="/$(arg uav_name)/uvdar/uvPointsSeen"/>
      <remap from="~blinkersSeen" to="/$(arg uav_name)/uvdar/blinkersSeen"/>
      <remap from="~estimatedFramerate" to="/$(arg uav_name)/uvdar/estimatedFramerate"/>
      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw"/>


      </node>
      
    <!-- <node name="uvdar_uav_pose_calculator_node" pkg="uvdar" type="uvdar_uav_pose_calculator_node" output="screen" launch-prefix="debug_roslaunch"> --> 
    <node name="uvdar_uav_pose_calculator_node" pkg="uvdar" type="uvdar_uav_pose_calculator_node" output="screen"> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>

      <param name="debug"		type="bool" value="false"/> 
      <param name="gui"             type="bool" value="$(arg gui)"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="frequencies_per_target" type="int" value="2"/>
      <param name="calib_file" type="string" value="$(arg calib_file_uv)"/>
      <param name="mask_file" type="string" value="$(arg mask_file)"/>

      <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3), $(arg frequency4)] </rosparam>


      <rosparam param="blinkers_seen_topics"> ["blinkersSeen"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimatedFramerate"] </rosparam>
      <rosparam param="camera_frames" subst_value="true"> ["uvcam_$(arg uav_name)"] </rosparam>

      <remap from="~blinkersSeen" to="/$(arg uav_name)/uvdar/blinkersSeen"/>
      <remap from="~estimatedFramerate" to="/$(arg uav_name)/uvdar/estimatedFramerate"/>
      <remap from="~measuredPose1" to="/$(arg uav_name)/uvdar/measuredPose1"/>
      <remap from="~measuredPose2" to="/$(arg uav_name)/uvdar/measuredPose2"/>
    </node>


    <!-- <node pkg="image_transport" type="republish" name="republish_bluefox" output="screen" args="compressed in:=bluefox/image_raw raw out:=bluefox2/image_raw"  /> -->
  
  <!-- <node pkg="uvdar" type="uvdar_kalman_node" name="filterUvdar" output="screen" launch-prefix="debug_roslaunch"> -->
  <!-- <node pkg="uvdar" type="uvdar_reprojector_node" name="reprojector_node" output="screen" launch-prefix="debug_roslaunch"> -->
  
  <!-- <node pkg="uvdar" type="uvdar_kalman_node" name="filterUvdar" output="screen" launch-prefix="debug_roslaunch"> -->
  <!-- <node pkg="uvdar" type="uvdar_kalman_node" name="filterUvdar" output="screen"> -->
  <!--   <param name="output_frame" value="$(arg frame_uvdar)"/> -->

  <!--   <remap from="measuredPose1" to="/$(arg uav_name)/uvdar/measuredPose1"/> -->
  <!--   <remap from="measuredPose2" to="/$(arg uav_name)/uvdar/measuredPose2"/> -->
  <!--   <remap from="filteredPose1" to="/$(arg uav_name)/uvdar/filteredPose1"/> -->
  <!--   <remap from="filteredPose2" to="/$(arg uav_name)/uvdar/filteredPose2"/> -->

  <!-- </node> -->
    </group>


  <!-- </group> -->



  </launch>
